core/src/main/scala/org/broadinstitute/dsde/rawls/googleProject/GoogleProjectRegistrationService.scala

```diff
@@ -0,0 +1,103 @@
+    new GoogleProjectRegistrationService(ctx,
+                                         dataSource,
+                                         samDAO,
+                                         googleProjectRegRepository,
+                                         billingRepository,
+                                         googleServicesDAO
+    )
+
+}
+
+class GoogleProjectRegistrationService(protected val ctx: RawlsRequestContext,
+                                       val dataSource: SlickDataSource,
+                                       val samDAO: SamDAO,
+                                       val googleProjectRegRepo: GoogleProjectRegistrationRepository,
+                                       val billingRepository: BillingRepository,
+                                       val googleServicesDAO: GoogleServicesDAO
+)(implicit
+  protected val executionContext: ExecutionContext
+) {
+
+  def registerGoogleProject(googleProjectReg: GoogleProjectRegistration): Future[GoogleProjectRegistration] =
+    for {
+      existingGoogleProjectReg <- googleProjectRegRepo.getGoogleProjectRegistration(googleProjectReg.googleProjectId)
+      billingProject <- billingRepository.getBillingProject(googleProjectReg.billingProjectId).map {
+        maybeBillingProject =>
+          maybeBillingProject.getOrElse(
+            throw new RawlsExceptionWithErrorReport(
+              errorReport = ErrorReport(
+                StatusCodes.NotFound,
+                "Billing project does not exist or you do not have permission to perform this action."
+              )
+            )
+          )
+      }
+      _ <- samDAO
+        .listUserActionsForResource(SamResourceTypeNames.billingProject, billingProject.projectName.value, ctx)
+        .map { actions =>
+          if (actions.isEmpty)
+            throw new RawlsExceptionWithErrorReport(
+              errorReport =
+                ErrorReport(StatusCodes.NotFound,
+                            "Billing project does not exist or you do not have permission to perform this action."
+                )
+            )
+          else if (!actions.contains(SamBillingProjectActions.link))
+            throw new RawlsExceptionWithErrorReport(
+              errorReport = ErrorReport(StatusCodes.Forbidden, "You do not have permission to perform this action.")
+            )
+        }
+      canLinkGoogleProject <- samDAO
+        .userHasAction(SamResourceTypeNames.googleProject,
+                       googleProjectReg.googleProjectId.value,
+                       SamGoogleProjectActions.link,
+                       ctx
+        )
+      _ = if (!canLinkGoogleProject)
+        throw new RawlsExceptionWithErrorReport(
+          errorReport = ErrorReport(StatusCodes.Forbidden, "You do not have permission to perform this action.")
+        )
+      result <- existingGoogleProjectReg match {
+        case Some(existing) if existing.billingProjectId == googleProjectReg.billingProjectId =>
```

Comment:
Do I need to check the billing account name as well, in case the billing project has changed billing accounts?

Reply:
Just to make sure I understand: On a register google project api request that includes (1) google project id and (2) rawls billing project uuid, we're checking to see if the rawls DB already has an entry matching both fields. If so, we return a success response. 

If that's correct, then I don't think we need to check the underlying google billing account. That would be an error state, but it should have been covered by the update billing endpoint.